core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -103,25 +109,11 @@ public void onComplete() {
                 topicPartitionData, shareFetchData.groupId(), shareFetchData.fetchParams());
 
         try {
-            Seq<Tuple2<TopicIdPartition, LogReadResult>> responseLogResult = replicaManager.readFromLog(
-                shareFetchData.fetchParams(),
-                CollectionConverters.asScala(
-                    topicPartitionData.entrySet().stream().map(entry ->
-                        new Tuple2<>(entry.getKey(), entry.getValue())).collect(Collectors.toList())
-                ),
-                QuotaFactory.UnboundedQuota$.MODULE$,
-                true);
-
-            Map<TopicIdPartition, FetchPartitionData> responseData = new HashMap<>();
-            responseLogResult.foreach(tpLogResult -> {
-                TopicIdPartition topicIdPartition = tpLogResult._1();
-                LogReadResult logResult = tpLogResult._2();
-                FetchPartitionData fetchPartitionData = logResult.toFetchPartitionData(false);
-                responseData.put(topicIdPartition, fetchPartitionData);
-                return BoxedUnit.UNIT;
-            });
-
-            log.trace("Data successfully retrieved by replica manager: {}", responseData);
+            Map<TopicIdPartition, FetchPartitionOffsetData> responseData;
+            if (logReadResponseFromTryComplete == null || logReadResponseFromTryComplete.isEmpty())
+                responseData = readFromLog(topicPartitionData);
+            else
+                responseData = combineLogReadResponseForOnComplete(topicPartitionData);
```

Comment:
Can it ever occur that you have non-empty `logReadResponseFromTryComplete` but `topicPartitionData` came from fresh aquisition from line 98 (topicPartitionData = acquirablePartitions();). I think never, can you just write comments for this.